File: 005_reverse_binary_tree/ReverseBinaryTree.java
class ReverseBinaryTree {
    public static class Node {
        public int value;
        public Node left;
        public Node right;

        public Node(int v) {
            value = v;
        }

        public String toString() {
            return String.format("Value: %d, Left: %d, Right: %d", 
                value, left.value, right.value);
        }
    }

    private static void solution1(Node root) {
        Node tmp = root.left;
        root.left = root.right;
        root.right = tmp;

        if (null != root.left) {
            solution1(root.left);
        }

        if (null != root.right) {
            solution1(root.right);
        }
    }

    public static void main(String[] args) {
        Node tree = new Node(1);
        tree.left = new Node(2);
        tree.right = new Node(3);
        solution1(tree);

        System.out.println(tree.toString());
    }
}

File: 005_reverse_binary_tree/ReverseBinaryTree.cpp
#include <iostream>
#include <string>

#include <cstdio>

using namespace std;

struct Node {
    int value;
    Node* left;
    Node* right;

    Node(int v): value(v) {};
    Node(int v, Node* l, Node* r): value(v), left(l), right(r) {};

    string toString() {
        int size = 100;
        char buffer[size];
        snprintf(buffer, size, "Value: %d, Left: %d, Right: %d", 
            value, left->value, right->value);
        return string(buffer);
    }
};

class BinaryTree {
public:
    BinaryTree(Node* node): root(node) {};
    ~BinaryTree() {
        destroyTree();
    };

    Node* getRoot() const {
        return root;
    }

    void destroyTree() {
        destroyTree(root);
    }

    string toString() {
        return root->toString();
    }
private:
    Node* root;

    void destroyTree(Node* node) {
        if (nullptr != node) {
            destroyTree(node->left);
            destroyTree(node->right);
            delete node;
        }
    }
};

/*
 * Recursively mutates the original tree
 */
void solution1(Node* root) {
    if (nullptr == root) {
        return;
    }

    Node* tmp = root->left;
    root->left = root->right;
    root->right = tmp;

    solution1(root->left);
    solution1(root->right);
}

int main() {
    Node* node = new Node(1);
    node->left = new Node(2);
    node->right = new Node(3);
    BinaryTree* tree = new BinaryTree(node);

    cout << tree->toString() << endl;

    solution1(tree->getRoot());
    cout << tree->toString() << endl;

    delete tree;

    return 0; 
}
